core/src/main/java/com/graphhopper/reader/dem/HeightTile.java

```diff
@@ -85,38 +95,37 @@ public double getHeight(double lat, double lon) {
         if (deltaLon > lonHigherBound || deltaLon < lowerBound)
             throw new IllegalStateException("longitude not in boundary of this file:" + lat + "," + lon + ", this:" + this.toString());
 
-        // first row in the file is the northernmost one
-        // http://gis.stackexchange.com/a/43756/9006
-        int lonSimilar = (int) (width / horizontalDegree * deltaLon);
-        // different fallback methods for lat and lon as we have different rounding (lon -> positive, lat -> negative)
-        if (lonSimilar >= width)
-            lonSimilar = width - 1;
-        int latSimilar = height - 1 - (int) (height / verticalDegree * deltaLat);
-        if (latSimilar < 0)
-            latSimilar = 0;
-
-        // always keep in mind factor 2 because of short value
-        int daPointer = 2 * (latSimilar * width + lonSimilar);
-        int value = heights.getShort(daPointer);
-        AtomicInteger counter = new AtomicInteger(1);
-        if (value == Short.MIN_VALUE)
-            return Double.NaN;
-
-        if (calcMean) {
-            if (lonSimilar > 0)
-                value += includePoint(daPointer - 2, counter);
-
-            if (lonSimilar < width - 1)
-                value += includePoint(daPointer + 2, counter);
-
-            if (latSimilar > 0)
-                value += includePoint(daPointer - 2 * width, counter);
-
-            if (latSimilar < height - 1)
-                value += includePoint(daPointer + 2 * width, counter);
+        double elevation;
+        if (interpolate) {
+            double x = (width - 1) * deltaLon / horizontalDegree;
+            double y = (height - 1) * (1 - deltaLat / verticalDegree);
+            int left = (int) x;
+            int top = (int) y;
+            int right = left + 1;
+            int bottom = top + 1;
+
+            double w00 = getHeightSample(left, top);
+            double w01 = getHeightSample(left, bottom);
+            double w10 = getHeightSample(right, top);
+            double w11 = getHeightSample(right, bottom);
+
+            double topEle = linearInterpolate(w00, w10, x - left);
+            double bottomEle = linearInterpolate(w01, w11, x - left);
+            elevation = linearInterpolate(topEle, bottomEle, y - top);
+        } else {
+            // first row in the file is the northernmost one
+            // http://gis.stackexchange.com/a/43756/9006
+            int x = (int) (width / horizontalDegree * deltaLon);
+            // different fallback methods for lat and lon as we have different rounding (lon -> positive, lat -> negative)
+            if (x >= width)
+                x = width - 1;
+            int y = height - 1 - (int) (height / verticalDegree * deltaLat);
+            if (y < 0)
+                y = 0;
+
+            elevation = getHeightSample(x, y);
         }
-
-        return (double) value / counter.get();
+        return elevation < MIN_ELEVATION ? Double.NaN : elevation;
```

Comment:
Instead of returning NaN should we better throw an exception? NaN will cause trouble somewhere else.

Reply:
This is the behavior from line 102-103 of the previous version of this code - NaN was the value returned for voids. I checked out PointList.calcDistance when elevation is NaN and it can't handle it. Is there code that I'm missing elsewhere that handles these voids?

Reply:
also FWIW valhalla does a similar range check when using mapzen/skadi terrain tiles https://github.com/valhalla/valhalla/blob/9dd8b88aa93b91888c4a6da8b09454a1ec9a0c2b/src/skadi/sample.cc#L224 and has some special case handling to skip over these when computing grade/total elevation gain/3d profiles with missing elevation data. Maybe this (or a separate) review should tighten up elevation void handling?

Reply:
I started working on a PR to handle voids downstream, and it started looking rather large so I opened up #1946 so we can discuss the best path forward.